### PR TITLE
Resolved problems with ssd_resnet34_mlperf_opset10

### PR DESCRIPTION
**Pull request by mbencer**
The following problems with mlperf ssd_resnet34_1200 were resolved (36674):
- removing TopK axis normalization from onnx importer (it is handled by nG op)
- relaxed Loop shape inference for output taken by `get_concatenated_slices` (only Concat axis must be dynamic if number of iterations is not determined during shape inference)
- fixed handling nodes from parent graph scope used in Loop's body during importing (it is scenario from sdd resnet model)
- added applying Squeeze on axis 0 for all concatenated outputs (it current version it was applied only for scalars, but it was confirmed that it is needed always)
- added handling `trip_count` passed as max(int64_t) in the same way as skipped `trip_count` (infinitive loop). It can improve Loop's shape inference

### CODE REVIEW

File: inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp [hunk 1]
```diff
@@ -317,11 +318,30 @@ CNNNetworkNGraphImpl::reshape(const std::map<std::string, std::vector<size_t>>&
...
+        return sub_functions;
+    }
+} // namespace
+
 void
 CNNNetworkNGraphImpl::reshape(const std::map<std::string, std::vector<size_t>>& inputShapes) {
     OV_ITT_SCOPED_TASK(itt::domains::IE, "CNNNetworkNGraphImpl::reshape");
 
     auto params = _ngraph_function->get_parameters();
+    auto sub_functions = get_all_sub_functions(_ngraph_function);
```
Comment (mbencer) [inline inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp:344]: I am not sure if it is an optimum solution. Maybe in order to avoid checking all nodes during reshape `Function` should have field `m_contain_sub_functions`? Or maybe there is some other possibility to replace parameters in sub-functions (SubGraphOp's bodies)

File: ngraph/test/util/engine/ie_engines.cpp [hunk 2]
```diff
@@ -126,12 +126,16 @@ namespace
...
+    auto cnn_network = InferenceEngine::CNNNetwork(m_function);
+
+    auto shapes = cnn_network.getInputShapes();
+    shapes["image:0"] = InferenceEngine::SizeVector{1, 1200, 1200, 3};
+    cnn_network.reshape(shapes);
+
     m_network_inputs = cnn_network.getInputsInfo();
     m_network_outputs = cnn_network.getOutputsInfo();
-
-    InferenceEngine::Core ie;
```
Comment (lazarevevgeny) [inline ngraph/test/util/engine/ie_engines.cpp:134]: Is it a debug code?

File: ngraph/test/onnx/onnx_import_controlflow.in.cpp [hunk 3]
```diff
@@ -439,3 +456,29 @@ NGRAPH_TEST(${BACKEND_NAME}, onnx_controlflow_loop_power)
...
+        "ssd_mobilenet_v1_coco_2018_01_28.onnx");
+}
+
+NGRAPH_TEST(${BACKEND_NAME}, test_ssd)
+{
+    const auto function = onnx_import::import_onnx_model(
+        "/home/mbencer/workspace/models_to_loop_test/mlperf_ssd_resnet34_1200/"
+        "ssd_resnet34_mAP_20.2.onnx");
+
+    const std::string base_path = "/home/mbencer/workspace/ssd";
```
Comment (lazarevevgeny) [inline ngraph/test/onnx/onnx_import_controlflow.in.cpp:473]: Hardcoded paths. Please, fix

File: inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp [hunk 4]
```diff
@@ -331,7 +351,18 @@ CNNNetworkNGraphImpl::reshape(const std::map<std::string, std::vector<size_t>>&
         auto newParam = std::make_shared<::ngraph::op::Parameter>(param->get_element_type(), shape);
         newParam->set_friendly_name(param->get_friendly_name());
         _ngraph_function->replace_parameter(i, newParam);
+
+        // Update parameters in sub-functions
```
Comment (lazarevevgeny) [inline inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp:355]: Please, describe in the comments why this change is needed.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: ngraph/test/onnx/onnx_import_controlflow.in.cpp (hunk 3 above)
Comment (mbencer) [reply ngraph/test/onnx/onnx_import_controlflow.in.cpp:473]: debug code ;) removed

File: ngraph/test/util/engine/ie_engines.cpp (hunk 2 above)
Comment (mbencer) [reply ngraph/test/util/engine/ie_engines.cpp]: right, debug code ;) removed

File: ngraph/core/src/op/loop.cpp [hunk 5]
```diff
@@ -19,6 +19,7 @@
 #include "ngraph/factory.hpp"
 #include "ngraph/graph_util.hpp"
 #include "ngraph/opsets/opset5.hpp"
+#include "ngraph/pass/visualize_tree.hpp"
```
Comment (tsocha) [inline ngraph/core/src/op/loop.cpp:22]:
```suggestion
```

**Review by tsocha — CHANGES_REQUESTED** [review]

**Review by tomdol — APPROVED** [review]

File: inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp [hunk 6]
```diff
@@ -331,7 +351,18 @@ CNNNetworkNGraphImpl::reshape(const std::map<std::string, std::vector<size_t>>&
...
+        for (auto& sub_func : sub_functions) {
+            const auto sub_func_params = sub_func->get_parameters();
+            for (auto paramt_it = sub_func_params.begin(); paramt_it != sub_func_params.end(); ++paramt_it) {
+                if ((*paramt_it)->get_friendly_name() == newParam->get_friendly_name()) {
```
Comment (ilyachur) [inline inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp:359]: I never thought that user can change the subgraph parameters (subgraphs exist in TensorIterator and Loop operations). Why do we need it? I am not sure that it is correct.

File: inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp (hunk 4 above)
Comment (mbencer) [reply inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp]: @lazarevevgeny @itikhono It was solution of secondary problem. The root cause was incorrect handling nodes from parent scope used in Loop body during importing. Changes from `cnn_network_ngraph_impl.cpp` were reverted.

File: ngraph/core/src/op/loop.cpp (hunk 5 above)
Comment (mbencer) [reply ngraph/core/src/op/loop.cpp]: done

File: inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp (hunk 6 above)
Comment (mbencer) [reply inference-engine/src/inference_engine/cnn_network_ngraph_impl.cpp]: ^^as describe above it was solution to the secondary problem. Reverted.

**Review by tsocha — APPROVED** [review]

File: ngraph/python/tests/__init__.py [hunk 7]
```diff
@@ -137,10 +137,12 @@ def xfail_test(reason="Mark the test as expected to fail", strict=True):
                                       "Argument element types are inconsistent.")
 xfail_issue_43742 = xfail_test(reason="RuntimeError: nGraph does not support the following ONNX operations:"
                                       "If")
-xfail_issue_43439 = xfail_test(reason="Check 'tensor_rank.is_static()' failed at "
-                                      "ngraph/core/src/validation_util.cpp:884:"
-                                      "map_1/while/select_bboxes/sort_bboxes_10/TopKV2 "
-                                      "Rank must be static in order to normalize negative axis=-1")
+xfail_issue_44602 = xfail_test(reason="While validating node 'v0::Squeeze"
```
Comment (lazarevevgeny) [inline ngraph/python/tests/__init__.py:140]: May be it make sense to find the root cause of 44602 and fix it in this PR also and remove the xfail?)

Comment (itikhono) [reply ngraph/python/tests/__init__.py:140]: I agree with Evgeny, it's better to fix the original problem. Looks like dynamic shapes appear somewhere in the network, it leads to issues in TopK op (fixed by relaxing shape inference of Loop op) and now in Squeeze op.

Comment (mbencer) [reply ngraph/python/tests/__init__.py:140]: Probably yes, but main intention was to avoid a large PR to review ;)

Comment (mbencer) [reply ngraph/python/tests/__init__.py:140]: ^^ `Squeeze` case is rather separate problem, `map_1/while/select_bboxes/sort_bboxes_1/TopKV2:1` is fully static `{400}` after this change

**Comment by gladilov-gleb** [comment]
@lazarevevgeny @itikhono @mbencer 

I suggest to merge https://github.com/openvinotoolkit/openvino/pull/3641 first since it fixes shape inference issues of Loop not only for its ouputs, but body inputs as well + add some additional tests. https://github.com/openvinotoolkit/openvino/pull/3641 will allow to continue work of Myriad-X team on dynamic models enablement.

**Comment by mbencer** [comment]
> @lazarevevgeny @itikhono @mbencer
> 
> I suggest to merge #3641 first since it fixes shape inference issues of Loop not only for its ouputs, but body inputs as well + add some additional tests. #3641 will allow to continue work of Myriad-X team on dynamic models enablement.

@gladilov-gleb #3641 covers all my cases so I synchronize my PR after #3641 merge to master

File: ngraph/python/tests/__init__.py (hunk 7 above)
Comment (mbencer) [reply ngraph/python/tests/__init__.py]: @lazarevevgeny @itikhono Problem with Squeeze exception was resolved. Unfortunately the model fails during execution because of dynamic shape Loop (we cannot determine number of iterations during shape inference because termination condition body output is not a constant). The problem should be resolved when CPU starts support dynamic shapes.

**Review by lazarevevgeny — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]
